drivers:platform:mbed: IRQ platform drivers implementation

mphalke opened the pull request:
1. Implemented the IRQ handling for External (GPIO), Ticker and UART Rx IRQs.
2. Callback registration and invokement implementaion is slightly different than the conventional
IRQ platform drivers. Dedicated callback is needed for each IRQ. Callback function in mbed
expects no argument and void return type.

**gpio irq example:**

```
// -------------------------- IRQ callback function ----------------------------------
void ext_int_callback(void *ctx, uint32_t event, void *extra)
{
}

// -------------------------- Platform specific ----------------------------------
/* External interrupt Mbed platform specific parameters */
struct mbed_irq_init_param mbed_ext_int_extra_init_params = {
	.int_mode = EXT_IRQ_RISE,
	.ext_int_pin = CONV_MON_GPIO,
	.int_obj_type = NULL
};

// -------------------------- Generic IRQ API stuff ----------------------------------
/* External interrupt generic init parameters */
static struct irq_init_param ext_int_init_params = {
	.irq_ctrl_id = 0,
	.extra = &mbed_ext_int_extra_init_params,
	.platform_ops = &irq_ops
};

/* External interrupt callback descriptor */
static struct callback_desc ext_int_callback_desc = {
	ext_int_callback,
	NULL,
	NULL,
};

/* External interrupt descriptor */
struct irq_ctrl_desc *external_int_desc;

/* Initialize IRQ */
if (irq_ctrl_init(&external_int_desc, &ext_int_init_params) != SUCCESS) {
	return FAILURE;
}

/* Register a callback function for external interrupt */
if (irq_register_callback(external_int_desc,
		  EXTERNAL_INT_ID1,
		  &ext_int_callback_desc) != SUCCESS) {
      return FAILURE;
}

/* Enable external interrupt */
if (irq_enable(external_int_desc, EXTERNAL_INT_ID1) != SUCCESS) {
	return FAILURE;
}
```

**gpio uart_rx example:**
```
// -------------------------- UART Specific ----------------------------------
/* Initialize the UART */
struct uart_desc *uart_desc;
uart_init(&uart_desc, &uart_init_params);

// -------------------------- IRQ callback function ----------------------------------
void uart_rx_callback(void *ctx, uint32_t event, void *extra)
{
}

// -------------------------- Platform specific ----------------------------------
/* UART Rx interrupt Mbed platform specific parameters */
struct mbed_irq_init_param mbed_uart_rx_int_extra_init_params = {
        .irq_ctrl_id = 0,
	.int_obj_type = (((struct mbed_uart_desc *)(uart_desc->extra))->uart_port);
};

// -------------------------- Generic IRQ API stuff ----------------------------------
/* UART Rx interrupt generic init parameters */
static struct irq_init_param uart_rx_int_init_params = {
	.irq_ctrl_id = UART_RX_INT_ID1,
	.extra = &mbed_uart_rx_int_extra_init_params,
	.platform_ops = &irq_ops
};

/* UART Rx interrupt callback descriptor */
static struct callback_desc uart_rx_int_callback_desc = {
	uart_rx_callback,
	NULL,
	NULL,
};

/* UART Rx interrupt descriptor */
struct irq_ctrl_desc *uart_rx_int_desc;

/* Initialize IRQ */
if (irq_ctrl_init(&uart_rx_int_desc, &uart_rx_int_init_params) != SUCCESS) {
	return FAILURE;
}

/* Register a callback function for external interrupt */
if (irq_register_callback(uart_rx_int_desc,
		  UART_RX_INT_ID1,
		  &uart_rx_int_callback_desc) != SUCCESS) {
      return FAILURE;
}

/* Enable external interrupt */
if (irq_enable(uart_rx_int_desc, UART_RX_INT_ID1) != SUCCESS) {
	return FAILURE;
}
```

Signed-off-by: mahphalke <Mahesh.Phalke@analog.com>

adrimbarean commented on drivers/platform/mbed/mbed_irq.cpp:
```diff
@@ -0,0 +1,569 @@
+* for every IRQ instance. Also, the 'callback_function' required for
+* mbed layer expects void argument and void return type.
+*/
+typedef struct {
```
We shouldn't use typedef in this case, according to the [Linux kernel coding style](https://www.kernel.org/doc/html/v4.11/process/coding-style.html#typedefs).

adrimbarean commented on drivers/platform/mbed/mbed_irq.cpp:
```diff
@@ -0,0 +1,569 @@
+* mbed layer expects void argument and void return type.
+*/
+typedef struct {
+	struct callback_desc callback_ext_int_id1;
+	struct callback_desc callback_ext_int_id2;
+	struct callback_desc callback_ext_int_id3;
+	struct callback_desc callback_ext_int_id4;
+	struct callback_desc callback_ext_int_id5;
```
We could make those an array `struct callback_desc callback_ext_int_id[5];` and refer to them by index. I think this will help with the double `switch` scenario in the register function.

mphalke replied:
Make sense. I have restructured this bit differently (using array of callback structure) to simplify some of the later code.

adrimbarean commented on drivers/platform/mbed/mbed_irq.cpp:
```diff
@@ -0,0 +1,569 @@
+	mbed_irq_desc->ext_int_pin = ((struct mbed_irq_init_param *)
+				      param->extra)->ext_int_pin;
+
+	switch (param->irq_ctrl_id) {
```
Here, the `irq_ctrl_id` member stands for the ID of the interrupt controller, not the ID of the interrupt. This is so that we can have multiple interrupt controllers in the same system, possibly cascading into each other. We should just save this in the `irq_desc` and do this `switch case` later, in the register interrupt method. Is it possible to do this?

mphalke replied:
Sorry not sure if I got your comment correctly, so just clarifying it.

Do you mean to say that ```irq_id``` passed as an argument to other functions in this module is not the same as ```irq_ctrl_id```? That means the ```irq_id enum``` value from ```mbed_irq.h``` file will be passed as argument to ```irq_id``` parameter for all other functions except this init function. The code under this ```switch()``` case would be then part of ```mbed_irq_register_callback()``` function.  
If this the case, then it make sense. I was also wondering why ```irq_id``` is passed again to all the functions if it is already passed to init function via irq init structure, but both seems to be different. 

Please confirm my understanding.

adrimbarean replied:
Yes, you are right. `irq_ctrl_id` is the ID of the interrupt controller that is used, and is set in the init function, and `irq_id` is the ID of the interrupt within the interrupt controller.

mphalke replied:
Thanks for the confirmation. I will implement the changes accordingly.

adrimbarean commented on drivers/platform/mbed/mbed_irq.cpp:
```diff
@@ -0,0 +1,569 @@
+
+	switch (irq_id) {
+	case  EXTERNAL_INT_ID1:
+		mbed_irq_callbacks.callback_ext_int_id1.callback = NULL;
```
We should call `ext_interrupt->rise/fall(NULL);` here to unregister in the DFP as well.

adrimbarean commented on drivers/platform/mbed/mbed_irq.cpp:
```diff
@@ -0,0 +1,569 @@
+
+	/* Unregister all callbacks */
+	for (irq_id = 0; irq_id < (uint8_t)NB_INTERRUPTS; irq_id++) {
+		if (mbed_irq_unregister(desc, irq_id) != SUCCESS)
```
Should we call `disable` here as well?

adrimbarean commented on drivers/platform/mbed/mbed_irq.cpp:
```diff
@@ -0,0 +1,569 @@
+
+	case TICKER_INT_ID:
+		/* Ticker doesn't have destructor implementation so only detach is used */
+		ticker = (mbed::Ticker *)(((struct mbed_irq_desc *)desc->extra)->int_obj);
+		ticker->detach();
+		break;
```
We should `delete(((struct mbed_irq_desc *)desc->extra)->int_obj);` here as well.

adrimbarean reviewed (commented):
Looks good overall. A few comments and questions.

mphalke commented:
V1:

- Created an array of structure for storing irq general callback structure parameters and mbed irq callback function.
- Removed the IRQ object creation from ```irq_ctr_init()``` function and moved into ```mbed_irq_callback_register()``` function as per review comment.
- Any attach (enabling) of interrupt is done from ```mbed_irq_enable()``` function as attaching interrupt automatically enabled it.
- Detach (disabling) of interrupt is also called from ```mbed_irq_unregister_callback()``` function to disable active interrupt callbacks.
- Other restructuring and code cleanup for better readability of code.

mphalke commented:
Hi All,
Any comments on the recent (V1) changes?

mphalke commented:
Hi,
Can we merge this PR if no further comments?

buha commented:
Also, it would help a lot to have a usage example of how to set up a gpio and a uart interrupt on mbed using the code in this PR. I think we can even better discuss on irq setup example than on the driver itself.

mphalke commented:
> Also, it would help a lot to have a usage example of how to set up a gpio and a uart interrupt on mbed using the code in this PR. I think we can even better discuss on irq setup example than on the driver itself.

@buha I have added usage example on how to setup GPIO and UART Rx interrupts on mbed platform with current implementation.

buha commented:
Ok, so in light of the irq.c/irq.h changes which hopefully will get accepted:

```
/* UART Rx interrupt callback descriptor */
static struct callback_desc uart_rx_int_callback_desc = {
	uart_rx_callback,
	NULL,                     // <----------------- unused
	NULL,                    // <----------------- unused
};
```
Given that the ctx and config members of callback_desc are not needed, I suggest you implement a platform specific struct mbed_callback that only holds the actual callback function. In my opinion, struct callback_desc should be abolished because it is impossible to have even the same callback signature (some platforms require void func(void), others void func(p1), others int func(p1, p2) etc.).

buha commented:
Also,

```

/* Register a callback function for external interrupt */
if (irq_register_callback(uart_rx_int_desc,
		  UART_RX_INT_ID1,                                             // <--------------- inq_id
		  &uart_rx_int_callback_desc) != SUCCESS) {
      return FAILURE;
}
```
irq_id should not be a custom defined enum, but it should be the actual interrupt id as taken from the interrupt controller documentation.

In this case, I think `enum irq_id` should be abolished, and we should try to use directly the interrupt ID as exported by the mbed libraries. These irq_* functions are intended to work as interrupt controller control functions, so consequently, the irq_id parameter should be precisely the interrupt number as the interrupt controller understands it on each platform.

buha commented on drivers/platform/mbed/mbed_irq.h:
```diff
@@ -0,0 +1,149 @@
+ *				.int_mode = EXT_IRQ_RISE
+ *			}
+ * */
+enum ext_irq_event {
```
In my opinion, this event should go together into the `struct mbed_callback` i mentioned above.
When you configure a GPIO interrupt, you may want to have two callbacks:
```
void rising_edge_irq(void) {
    // do something on rising edge
}

void falling_edge_irq(void) {
    // do something on falling edge
}
```
This should be configurable with `irq_register_callback(desc, irq_id, callback)`.
I think something like:
```
struct mbed_callback rising {
    .callback = rising_edge_irq,
    .event = MBED_RISING // invented macro name, look up the correct one in mbed libraries
}

struct mbed_callback falling {
    .callback = falling_edge_irq,
    .event = MBED_FALLING // invented macro name, look up the correct one in mbed libraries
}

irq_register_callback(desc, irq_id, rising)
irq_register_callback(desc, irq_id, falling)
```

mphalke commented:
> Ok, so in light of the irq.c/irq.h changes which hopefully will get accepted:
> 
> ```
> /* UART Rx interrupt callback descriptor */
> static struct callback_desc uart_rx_int_callback_desc = {
> 	uart_rx_callback,
> 	NULL,                     // <----------------- unused
> 	NULL,                    // <----------------- unused
> };
> ```
> 
> Given that the ctx and config members of callback_desc are not needed, I suggest you implement a platform specific struct mbed_callback that only holds the actual callback function. In my opinion, struct callback_desc should be abolished because it is impossible to have even the same callback signature (some platforms require void func(void), others void func(p1), others int func(p1, p2) etc.).

This make sense. If 'callback_desc' is abolished, things will be simpler and no need to define callback functions in the mbed_irq.c file as it is done currently. 
So I will wait for the other PR to get approved before doing this change.

mphalke commented:
> Also,
> 
> ```
> 
> /* Register a callback function for external interrupt */
> if (irq_register_callback(uart_rx_int_desc,
> 		  UART_RX_INT_ID1,                                             // <--------------- irq_id
> 		  &uart_rx_int_callback_desc) != SUCCESS) {
>       return FAILURE;
> }
> ```
> 
> irq_id should not be a custom defined enum, but it should be the actual interrupt id as taken from the interrupt controller documentation.
> 
> In this case, I think `enum irq_id` should be abolished, and we should try to use directly the interrupt ID as exported by the mbed libraries. These irq_* functions are intended to work as interrupt controller control functions, so consequently, the irq_id parameter should be precisely the interrupt number as the interrupt controller understands it on each platform.

The mbed library does not expose the interrupt controller ID/number for each interrupt. Mbed is an abstracted layer that can support multiple MCUs, so details of controller IDs, names, etc is hidden in the HAL layer (vendor specific) residing below Mbed layer, which we can't/shouldn't access directly. I understand other platform drivers are specific to certain MCU Or series of MCU, so it is easy to get interrupt ID for those.
